### PR TITLE
Allow viewing the log file for SelfDiagnostics for .NET Framework

### PR DESCRIPTION
**Pull request by xiang17**
Fixes Bug#2 in #1689. When running with .NET Framework, the log file cannot be opened as read-only, caused by default FileShare behavior in MemoryMappedFile in .NET Framework.

.NET Framework [opens the file in FileShare.None mode](https://referencesource.microsoft.com/#system.core/System/IO/MemoryMappedFiles/MemoryMappedFile.cs,148) but .NET Core [opens the file in FileShare.Read mode](https://github.com/dotnet/runtime/blob/master/src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs#L152).

## Changes

Use the [FileStream version](https://docs.microsoft.com/en-us/dotnet/api/system.io.memorymappedfiles.memorymappedfile.createfromfile?view=net-5.0#System_IO_MemoryMappedFiles_MemoryMappedFile_CreateFromFile_System_IO_FileStream_System_String_System_Int64_System_IO_MemoryMappedFiles_MemoryMappedFileAccess_System_IO_HandleInheritability_System_Boolean_) of `MemoryMappedFile.CreateFromFile` instead of the [string version](https://docs.microsoft.com/en-us/dotnet/api/system.io.memorymappedfiles.memorymappedfile.createfromfile?view=net-5.0#System_IO_MemoryMappedFiles_MemoryMappedFile_CreateFromFile_System_String_System_IO_FileMode_System_String_System_Int64_), because the the string version opens the file in different FileShare mode for .NET Framework and .NET Core.

Tested manually:
1. Modify code. Wrap `using (var activity = MyActivitySource.StartActivity("SayHello"))` block in [docs/trace/getting-started/Program.cs](https://github.com/open-telemetry/opentelemetry-dotnet/blob/master/docs/trace/getting-started/Program.cs#L34-L39) with `while (true)` loop.
2. Create a configurations file and run the project according to issue #1689.
3. When the program is running, open the output file (with name like "getting-started.exe.<ProcessId>.log"). The file should open successfully.

### CODE REVIEW

**Comment by xiang17** [comment]
I also ran the unit tests for the code before the fix. The test cases all passed for .NET Core but failed for .NET Framework.

OpenTelemetry - Microsoft Visual Studio 1_13_2021 4_37_15 PM - Copy (2)

File: src/OpenTelemetry/Internal/SelfDiagnosticsConfigRefresher.cs [hunk 1]
```diff
@@ -203,7 +210,40 @@ private void OpenLogFile(string newLogDirectory, int newFileSize)
...
+                // .NET Framework and .NET Core, here I am using the [FileStream version][2] of it.
+                // Taking the last four prameter values from [.NET Framework]
+                // (https://referencesource.microsoft.com/#system.core/System/IO/MemoryMappedFiles/MemoryMappedFile.cs,148)
+                // and [.NET Core]
+                // (https://github.com/dotnet/runtime/blob/master/src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs#L152)
+                // The parameter for FileAccess is different in type but the same in rules, both are Read and Write.
+                // The parameter for FileShare is different in values and in behavior.
+                // .NET Framework doesn't allow sharing but .NET Core allows reading by other programs.
+                // The last two parameters are the same values for both frameworks.
+                // [1]: https://docs.microsoft.com/en-us/dotnet/api/system.io.memorymappedfiles.memorymappedfile.createfromfile?view=net-5.0#System_IO_MemoryMappedFiles_MemoryMappedFile_CreateFromFile_System_String_System_IO_FileMode_System_String_System_Int64_
```
Comment (cijothomas) [inline src/OpenTelemetry/Internal/SelfDiagnosticsConfigRefresher.cs:224]: Please remove "en-us" from doc links.

File: test/OpenTelemetry.Tests/Internal/SelfDiagnosticsConfigRefresherTest.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            CreateConfigFile();
+            using var configRefresher = new SelfDiagnosticsConfigRefresher();
+
+            var outputFileName = Path.GetFileName(Process.GetCurrentProcess().MainModule.FileName) + "."
+                    + Process.GetCurrentProcess().Id + ".log";
+            var outputFilePath = Path.Combine(".", outputFileName);
+            using var file = File.Open(outputFilePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
+            byte[] actualBytes = new byte[MessageOnNewFile.Length];
+            file.Read(actualBytes, 0, actualBytes.Length);
+            Assert.Equal(MessageOnNewFile, actualBytes);
```
Comment (cijothomas) [inline test/OpenTelemetry.Tests/Internal/SelfDiagnosticsConfigRefresherTest.cs:42]: Do a cleanup of the created ConfigFile in finally?

**Comment by cijothomas** [comment]
please add this to changelog as well.

File: test/OpenTelemetry.Tests/Internal/SelfDiagnosticsConfigRefresherTest.cs (hunk 2 above)
Comment (xiang17) [reply test/OpenTelemetry.Tests/Internal/SelfDiagnosticsConfigRefresherTest.cs]: Done.

File: src/OpenTelemetry/Internal/SelfDiagnosticsConfigRefresher.cs (hunk 1 above)
Comment (xiang17) [reply src/OpenTelemetry/Internal/SelfDiagnosticsConfigRefresher.cs]: Done.

**Comment by xiang17** [comment]
> 
> 
> please add this to changelog as well.

Added.

**Comment by cijothomas** [comment]
@xiang17 please resolve confloicts. change look good for me.

**Comment by xiang17** [comment]
> 
> 
> @xiang17 please resolve confloicts. change look good for me.

Done.

**Review by cijothomas — APPROVED** [review]

**Pull request merged**